snapcraft/extensions/env_injector.py

```diff
@@ -0,0 +1,135 @@
+    
+    .. code-block:: shell
+        sudo snap set <snap-name> env.<key>=<value>
+
+    - To set environment variables for a specific application **inside** the snap:
+
+    .. code-block:: shell
+        sudo snap set <snap-name> apps.<app-name>.env.<key>=<value>
+
+    - To set environment file inside the snap:
+
+    .. code-block:: shell
+        sudo snap set <snap-name> env-file=<path-to-env-file>
+
+    """
+
+    @staticmethod
+    @overrides
+    def get_supported_bases() -> Tuple[str, ...]:
+        return ("core24",)
+
+
+    @staticmethod
+    @overrides
+    def get_supported_confinement() -> Tuple[str, ...]:
+        return ("strict", "devmode", "classic")
+
+
+    @staticmethod
+    @overrides
+    def is_experimental(base: Optional[str]) -> bool:
+        return True
+    
+    @overrides
+    def get_root_snippet(self) -> Dict[str, Any]:
+        return {}
+
+    
+    @overrides
+    def get_app_snippet(self, *, app_name: str) -> Dict[str, Any]:
+        """Return the app snippet to apply."""
+        return {
+            "command-chain": [ "bin/command-chain/env-exporter"],
+            "environment": {
+                "env_alias": f"{app_name}",
+            }
+        }
+
+    @overrides
+    def get_part_snippet(self, *, plugin_name: str) -> Dict[str, Any]:
+        return {
+             "build-environment": [
+                 { "VARS_INJECT_EXTENSION": "true" }
+             ]
+        }
+
+    @overrides
+    def get_parts_snippet(self) -> Dict[str, Any]:
+        toolchain = self.get_toolchain()
+        return {
+            f"env-injector/env-injector": {
+                "source": f"{get_extensions_data_dir()}/env-injector",
+                "plugin": "nil",
+                "build-snaps": [
+                    f"rustup",
+                ],
+                "build-packages": [
+                    f"musl-tools",
+                    f"upx-ucl", # for binary compression
+                ],
+                "override-build": f"""
+
+      rustup default stable
+      rustup target add {toolchain}
+
+      cargo build --target {toolchain} --release
+      mkdir -p $SNAPCRAFT_PART_INSTALL/bin/command-chain
+
+      cp target/{toolchain}/release/env-exporter $SNAPCRAFT_PART_INSTALL/bin/command-chain
+
+      # shrink the binary
+      upx --best --lzma target/{toolchain}/release/env-exporter
+      cp target/{toolchain}/release/env-exporter $SNAPCRAFT_PART_INSTALL/bin/command-chain/env-exporter-upx
+
+                """,
+
+            }
+        }
+
+    def get_toolchain(self):
+        # Dictionary mapping architecture names
+        toolchain = {
+            'amd64': 'x86_64-unknown-linux-musl',
+            'arm64': 'aarch64-unknown-linux-musl',
+            'armhf': 'arm-unknown-linux-musleabihf',
+            'riscv64': 'riscv64gc-unknown-linux-musl',
+            'ppc64el': 'powerpc64-unknown-linux-gnu',
+            's390x': 's390x-unknown-linux-musl',
+            'i386': 'i586-unknown-linux-musl',
```

Comment:
I'm not sure how reliable are these builds. The targets are taken from [Tier 2](https://doc.rust-lang.org/rustc/platform-support.html#tier-2-with-host-tools) which are "guaranteed to build". Why not chose from [Tier 1](https://doc.rust-lang.org/rustc/platform-support.html#tier-1-with-host-tools) that are "guaranteed to work" targets?

Reply:
In order to static linking, we need the toolchains with the `*-musl` prefix, the tier 1 has two problems: It only have `*-gnu` toolchains and only have support for `amd64` and `arm64`, also the tendency is that the Tier 2 toolchains became each time more realiable over the time, if you see [the list of toolchains](https://rust-lang.github.io/rustup-components-history/) there is already a Tier 2.5 which the documentation doesn't mention. In other words: We have to work with the tools that we have.

Reply:
Actually, there is another way to do static linking using gnu based toolchains, but it doesn't get close to the size of  when using the `musl` ones because of the size cost when static linking glibc, [please take a look at this analysis](https://gist.github.com/locnnil/c6f7dad389444ded0928bfc4c6f027aa).